File: client/internal/tools/patch_file.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// PatchFileTool performs a targeted find-and-replace edit on an existing file.
type PatchFileTool struct{}

type patchFileArgs struct {
	Path      string `json:"path"`
	OldString string `json:"old_string"`
	NewString string `json:"new_string"`
}

func (t *PatchFileTool) Name() string { return "patch_file" }

func (t *PatchFileTool) Description() string {
	return "Edit an existing file by replacing a specific string. Provide the exact text to find (old_string) and its replacement (new_string). The old_string must match exactly one location in the file. Use file_read first to see the current content."
}

func (t *PatchFileTool) Parameters() json.RawMessage {
	return Schema{
		Type: "object",
		Properties: map[string]SchemaProperty{
			"path":       {Type: "string", Description: "Path to the file to edit"},
			"old_string": {Type: "string", Description: "The exact text to find in the file (must match exactly once)"},
			"new_string": {Type: "string", Description: "The text to replace old_string with"},
		},
		Required: []string{"path", "old_string", "new_string"},
	}.MustMarshal()
}

func (t *PatchFileTool) Execute(_ context.Context, arguments string) (*ToolResult, error) {
	var args patchFileArgs
	if err := json.Unmarshal([]byte(arguments), &args); err != nil {
		return ErrorResult(fmt.Sprintf("invalid arguments: %v", err)), nil
	}
	if args.Path == "" {
		return ErrorResult("path is required"), nil
	}
	if args.OldString == "" {
		return ErrorResult("old_string is required"), nil
	}
	if args.OldString == args.NewString {
		return ErrorResult("old_string and new_string are identical — nothing to change"), nil
	}

	content, err := os.ReadFile(args.Path)
	if err != nil {
		if os.IsNotExist(err) {
			return ErrorResult(fmt.Sprintf("file not found: %s", args.Path)), nil
		}
		return ErrorResult(fmt.Sprintf("failed to read file: %v", err)), nil
	}

	fileStr := string(content)
	count := strings.Count(fileStr, args.OldString)

	switch count {
	case 0:
		// Show a snippet of the file to help the LLM correct its call
		snippet := fileStr
		if len(snippet) > 500 {
			snippet = snippet[:500] + "\n...(truncated)"
		}
		return ErrorResult(fmt.Sprintf("old_string not found in %s. File content starts with:\n%s", args.Path, snippet)), nil

	case 1:
		// Exactly one match — perform the replacement
		newContent := strings.Replace(fileStr, args.OldString, args.NewString, 1)
		if err := os.WriteFile(args.Path, []byte(newContent), 0644); err != nil {
			return ErrorResult(fmt.Sprintf("failed to write file: %v", err)), nil
		}
		return &ToolResult{Output: fmt.Sprintf("Replaced %d characters with %d characters in %s",
			len(args.OldString), len(args.NewString), args.Path)}, nil

	default:
		return ErrorResult(fmt.Sprintf("old_string matches %d locations in %s. Include more surrounding context in old_string to make it unique.", count, args.Path)), nil
	}
}

File: clients/vscode/src/tools/patchFile.ts
import * as vscode from 'vscode';
import { resolvePath } from './paths';
import { err, ok, parseArgs, ToolImpl } from './types';

interface PatchFileArgs {
  path: string;
  old_string: string;
  new_string: string;
}

const SNIPPET_BYTES = 500;

/**
 * patch_file shim — find-and-replace edit. old_string must match exactly
 * once. Mirrors the Go tool's error messages so the model gets the same
 * hints when matching fails. Applies via WorkspaceEdit so VS Code shows
 * the change inline and undo restores the prior state.
 */
export const patchFile: ToolImpl = async (raw, ctx) => {
  const parsed = parseArgs<PatchFileArgs>(raw);
  if ('error' in parsed) {
    return err(parsed.error);
  }
  const { path: rawPath, old_string: oldString, new_string: newString } = parsed.args;
  if (!rawPath) {
    return err('path is required');
  }
  if (!oldString) {
    return err('old_string is required');
  }
  if (oldString === newString) {
    return err('old_string and new_string are identical — nothing to change');
  }

  const uri = resolvePath(rawPath, ctx.workspaceRoot);
  let doc: vscode.TextDocument;
  try {
    doc = await vscode.workspace.openTextDocument(uri);
  } catch (e) {
    if ((e as { code?: string }).code === 'FileNotFound') {
      return err(`file not found: ${rawPath}`);
    }

    return err(`failed to read file: ${(e as Error).message}`);
  }

  const text = doc.getText();
  const matches = countOccurrences(text, oldString);
  if (matches === 0) {
    const snippet = text.length > SNIPPET_BYTES ? text.slice(0, SNIPPET_BYTES) + '\n...(truncated)' : text;

    return err(
      `old_string not found in ${rawPath}. File content starts with:\n${snippet}`,
    );
  }
  if (matches > 1) {
    return err(
      `old_string matches ${matches} locations in ${rawPath}. Include more surrounding context in old_string to make it unique.`,
    );
  }

  const startOffset = text.indexOf(oldString);
  const range = new vscode.Range(
    doc.positionAt(startOffset),
    doc.positionAt(startOffset + oldString.length),
  );

  const edit = new vscode.WorkspaceEdit();
  edit.replace(uri, range, newString);
  const applied = await vscode.workspace.applyEdit(edit);
  if (!applied) {
    return err('VS Code declined to apply the edit');
  }
  try {
    if (doc.isDirty) {
      await doc.save();
    }
  } catch {
    // best-effort
  }

  return ok(
    `Replaced ${oldString.length} chars with ${newString.length} chars in ${rawPath}`,
  );
};

function countOccurrences(haystack: string, needle: string): number {
  if (!needle) {
    return 0;
  }
  let count = 0;
  let from = 0;
  while (true) {
    const i = haystack.indexOf(needle, from);
    if (i < 0) {
      break;
    }
    count++;
    from = i + needle.length;
  }

  return count;
}
